lib/src/database.dart

```diff
@@ -152,6 +152,39 @@ class Query {
         .then((snapshot) => new DataSnapshot(snapshot));
   }
 
+  /// Listens for data changes at a particular location.
+  ///
+  /// This is the primary way to read data from a [Database]. Your callback will
+  /// be triggered for the initial data and again whenever the data changes.
+  /// Use [off] to stop receiving updates.
+  ///
+  /// return function to unsubscribe
+  Function on<T>(
+      String eventType, Function(DataSnapshot<T> snapshot) callback) {
+    var fn = allowInterop((snapshot) => callback(new DataSnapshot(snapshot)));
+    nativeInstance.on(eventType, fn);
+    return () => nativeInstance.off(eventType, fn);
```

Comment:
I like the way it looks here, but we are stepping away from the official API, plus we'll not be able to introduce the `off` method this way.

What if we introduce `QuerySubscription` class which stores reference to `fn` and return an instance of this class here? Something like this:

```dart
class QuerySubscription {
  final String eventType;
  final js.Query _nativeQuery;
  final Function _callback;

  // Private constructor to prevent from external use.
  QuerySubscription._(this.eventType, this._nativeQuery, this._callback);

  /// Cancels this subscription.
  void cancel() {
    _nativeQuery.off(_eventType, _callback);
  }

class Query {
  QuerySubscription on(String eventType, callback, [cancelCallback]) {
    // ...
    return QuerySubscription(eventType, nativeInstance, fn);
  }
  void off(QuerySubscription subscription) {
    subscription.cancel();
  }
}
```

We'd keep the `on` and `off` methods in place for easier discoverability. The new class is a slight departure from original interface but seems much more flexible this way. Thoughts?

Reply:
We should keep eventType in .off, since there are 3 cases of using .off 

> If a callback is not specified, all callbacks for the specified eventType will be removed. Similarly, if no eventType or callback is specified, all callbacks for the Reference will be removed.

so maybe:
```
class Query {
  void off(String eventType, QuerySubscription subscription) {
    subscription.cancel();
  }
//...
}
```

Reply:
I think we should keep using non-named parameters, to keep it similar to other methods in the class/lib, as well as the library. Dart is pretty type-strict and AOT, so it won't be possible to mistake the second parameter with a Function, right? 

I also don't think we need the assert check. In case that eventType and Function subscription not exists, firebase would just ignore the method call.

Another approach to pass the Function as 2nd parameter is to use a Map<Function, Function> to store the Function reference and it's allowInterop() return. I haven't tried though. Still pretty new to Dart.

Reply:
> Using original JS signature here is definitely very straightforward, but it is not something I'd ever write in Dart. Additionally it would require users to pass null for eventType if they pass a subscription, e.g. query.off(null, mySubscription) - a very weird pattern to use, which can also be misleading (what if user passed both but eventType in subscription differs from the eventType passed as an argument?).

We could assert for when `eventType == null` to restrict user from doing it that way. If the user want to cancel the `subscription` by only using it, he/she can just use `subscription.cancel()` , else we are using .off like in JS API. So there will be a total of 4 use cases,

What I meant is the unity of all the methods in a class rather than the original JS signature. If the rest was also using named params, I would vote for named params for .off() too :)

Btw, does that mean the named parameters will prevent the `dynamic` bypass?

Reply:
If that's the case, instead of `dynamic` eventTypeOrSubscription, I suggest we could split the .off method into three methods
```dart
class Query {
  void offAll(){
    nativeInstance.off();
  }
  void offEvent(String eventType){
    nativeInstance.off(eventType);
  }
  void offSubscription(QuerySubscription subscription){
      nativeInstance.off(eventType, subscription._callback);
  }
  
}
```

Generally, I find Union Type or dynamic type parameters quite easy to cause confusion. It would show up in popup suggestion all three use-cases when typed 'off'. Also, this solution would completely remove .off() method from Query class since it could not be the same anyway. It helps the users (maybe) to avoid unintended un-subscription. I only discover that .off method has 2 other use cases recently, before that I have always used .off(eventType, callback).

What do you think?

Reply:
Actually, how about just leaving subscription argument out so that we only have:

```dart
void off([String eventType]) {
  if (eventType == null) {
    nativeInstance.off();
  } else {
    nativeInstance.off(eventType);
  }
}
```

If the user already provides us an instance of `subscription` they might as well just call `subscription.cancel()`.

Reply:
that works. i'll fix that up